src/ecdsa/ellipticcurve.py

```diff
@@ -174,20 +175,29 @@ def __init__(self, curve, x, y, z, order=None, generator=False):
             self.__y = y
             self.__z = z
             self.__order = order
+        self.__generator = generator
         self.__precompute = []
-        if generator:
+
+    def _maybe_precompute(self):
+        if self.__generator and not self.__precompute:
+            order = self.__order
             assert order
-            i = 1
-            order *= 2
-            doubler = PointJacobi(curve, x, y, z, order)
-            order *= 2
-            self.__precompute.append((doubler.x(), doubler.y()))
-
-            while i < order:
-                i *= 2
-                doubler = doubler.double().scale()
+            try:
+                self._scale_lock.writer_acquire()
```

Comment:
and what if multiple threads execute this line at the same time?

I'm also not a fan of overloading the `_scale_lock`, it's used for a different purpose

Reply:
> and what if multiple threads execute this line at the same time?

it will block in https://docs.python.org/3.8/library/threading.html#threading.Lock.acquire

a deadlock would arise if that code blocking would have previously acquired a write lock upon which the precomputing thread would have to wait subsequently. 

anyways, we'll use our fork for now, works

Reply:
what I mean is that both threads will then add values to the precompute list, that's wrong

Reply:
no, because of `not self.__precompute` on line 182. or are you saying that this latter access has to be under the critical section as well

Reply:
.. but that would escalate each and every "maybe_precompute" call, even when already cached. another option would be to optain a read lock already for the initial access in "maybe_precompute", and then escalate that to write lock without releasing before. not sure if the lock wrapper in use even allows that. mmh. locks;) don't use them ..